test/unit/examples.test.ts

```diff
@@ -129,4 +130,113 @@ describe('Examples', () => {
             ],
         );
     });
+
+    /**
+     * @see https://hapi.dev/family/joi/api/?v=15.1.1#extendextension
+     */
+    it('Extending Joi for custom validation', () => {
+        // Custom validation functions must be added by using Joi's "extend" mechanism.
+
+        // These are utility types you may find useful to replace the return of a function.
+        // These types are derived from: https://stackoverflow.com/a/50014868
+        type ArgumentTypes<T> = T extends (...args: infer U ) => infer _R ? U: never;
+        type ReplaceReturnType<T, TNewReturn> = (...a: ArgumentTypes<T>) => TNewReturn;
+
+        // We are going to create a new instance of Joi, with our extended functionality: a custom validation
+        //  function that checks if each character in a string has "alternating case" (that is, each character has a
+        //  case different to those either side of it).
+
+        // For our own peace of mind, we're first going to update the type of the Joi instance to include our new
+        //  schema.
+
+        interface ExtendedStringSchema extends StringSchema {
+            alternatingCase(): this; // We're adding this method, only for string schemas.
+        }
+
+        type IJoi = typeof Joi; // Need to alias this, because `interface Foo extends typeof Joi` doesn't work.
+        interface CustomJoi extends IJoi {
+            // This allows us to use our extended string schema, in place of Joi's original StringSchema.
+            // E.g. instead of `Joi.string()` returning `StringSchema`, it now returns `ExtendedStringSchema`.
+            string: ReplaceReturnType<IJoi['string'], ExtendedStringSchema>;
+        }
```

Comment:
nit: Just to avoid the prefixing types with `I` can we call the type `OriginalJoi`?

```suggestion
        // Need to alias this, because `interface Foo extends typeof Joi` doesn't work.
        type OriginalJoi = typeof Joi;

        interface CustomJoi extends OriginalJoi {
            // This allows us to use our extended string schema, in place of Joi's original StringSchema.
            // E.g. instead of `Joi.string()` returning `StringSchema`, it now returns `ExtendedStringSchema`.
            string: ReplaceReturnType<OriginalJoi['string'], ExtendedStringSchema>;
        }
```

Reply:
Sounds good. 👍